### ISSUE TITLE
Compiler or analyzer should warn for lower-cased type names

### ISSUE THREAD

**Issue by jcouv** [dotnet/roslyn#56653]
During LDM discussion on records, there was interest to have an analyzer warn on lower-cased type names. The motivation is that knowing that we discourage lower-cased type names would allow us to introduce new keywords (like `record`) with less concern about the breaking change.

I'm not sure whether this should be done by the compiler, a roslyn analyzer or some BCL analyzer.

Design:
- We use `char.IsLower(...)` for lower-case check.
- `@` prefix avoids warning.
- All named types, included delegate types, are checked.
- Type and namespace aliases are checked too.

FYI @mikadumont @jmarolf

**Comment by CyrusNajmabadi** [comment]
If this is about the language and our ability to add such types in a more confident manner, then i think this should be part of the compiler proper.  That way every gets it, and everyone sees that this is the official guidance from the team on what you should/shouldn't do.

**Comment by jcouv** [comment]
Just realized, one issue we'll have to figure out is how we define "lower-case".

**Comment by CyrusNajmabadi** [comment]
All(char.IsLower)?

**Comment by jmarolf** [comment]
Agree with @CyrusNajmabadi that the _ideal_ here is for the compiler to emit a new warning as part of warning wave 7. If that cannot happen then we can consider shipping it as an analyzer. Considering that this is a fundamental part of C# it would seem to be appropriate for the dotnet/roslyn repo since dotnet/roslyn-analyzers is for warnings about .NET libraries instead of core language concepts.

**Comment by jcouv** [comment]
@jaredpar Should we contemplate doing this as part of warning wave **6** (ie. in .NET 6)?

**Comment by Youssef1313** [comment]
Will prefixing with `@` remove the warning?

Otherwise, I can't think of a good workaround, for example, for Xamarin's `System.nuint`:

![image](https://user-images.githubusercontent.com/31348972/134802828-54a22273-edf8-45ca-99e7-2f88f16f692f.png)

**Comment by jcouv** [comment]
From discussion with Jared, it's too late for .NET 6. We could make this any early preview for release after that (such as 17.1).

@Youssef1313 That's an excellent suggestion. It would make sense given that the `@` prefix allows you to use other keywords as identifiers.

**Comment by JoeRobich** [comment]
Design Meeting Notes:
- Can this be covered by the NamingStyleAnalyzer?
  - By default these diagnostics are suggestions and compiler would prefer warnings.
  - Naming styles are configurable which is at odds with what the compiler is asking for.

Proposal: Keep IDE analyzers as they are. The compiler will add a diagnostic behind a warning wave for VS 17.1 Preview 1.

**Comment by 333fred** [comment]
Notes: https://github.com/dotnet/csharplang/blob/main/meetings/2021/LDM-2021-10-13.md#warning-on-lowercase-type-names

**Comment by xoofx** [comment]
Hello there,

In Unity, we are using lower-case names for some special types that we consider as "primitives" (e.g [float4](https://github.com/Unity-Technologies/Unity.Mathematics/blob/master/src/Unity.Mathematics/float4.gen.cs))

So we have 2 questions:
- Will it be possible to disable this warning project wide?
- Does it apply only to declarations or If another project is consuming these names (e.g method with a parameter having these types), will propagate to users projects?

**Comment by xoofx** [comment]
Thanks for the response @CyrusNajmabadi, sorry I cross posted (thought I deleted this one above)

**Comment by WayneMV** [comment]
> All(char.IsLower)?

Will that work for languages like Arabic and Hebrew which don't differentiate letter case at all?

**Comment by 333fred** [comment]
> > All(char.IsLower)?
> 
> Will that work for languages like Arabic and Hebrew which don't differentiate letter case at all?

These cases will not warn, as we don't ever anticipate introducing keywords that use non-ascii characters.

**Comment by soc** [comment]
This is a great improvement. Is there already a recommended time table on which users can start replacing `int`, `float`, `string` and friends?

**Comment by 333fred** [comment]
> This is a great improvement. Is there already a recommended time table on which users can start replacing `int`, `float`, `string` and friends?

Replace them with what?

**Comment by CyrusNajmabadi** [comment]
Replace them with what?

**Comment by soc** [comment]
https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/built-in-types

**Comment by Youssef1313** [comment]
@soc Both alias C# keywords (e.g. `string`) and the .NET runtime types (e.g. `System.String`) works. Nothing has changed here. This is about user-defined types that are named in all lowercase. For example:

```csharp
class abc { } // warning
```

**Comment by soc** [comment]
The point is consistency.

**Comment by Youssef1313** [comment]
I'm not really sure what you mean.

**Comment by davidwengier** [comment]
I think they mean that if users aren't allowed to define a type that is all lowercase, then the runtime shouldn't either.

**Comment by Youssef1313** [comment]
If that's what @soc meant, then there are two things:

1. `string`, `int`, etc, are language keywords that map to PascalCase types. So they are not lowercase types.
2. Even if the runtime (or a regular user) wanted to define a lowercase type, you can always prefix it with `@`. For example, I'd imagine Xamarin will do that if they encountered the warning for their [nint](https://docs.microsoft.com/en-us/dotnet/api/system.nint?view=xamarin-ios-sdk-12) (and nuint).

**Comment by CyrusNajmabadi** [comment]
> @Youssef1313 The point is consistency: Do not make up rules for language users that you, as a language developer, don't intend to follow.

The point is exactly to have that block.  We *want* the language to be able to use these keywords without colliding with anything a user does.  If we continue to use the same space of names (like today), then nothing will change and we will have the same potential collision problem.

**Comment by soc** [comment]
@Youssef1313

1. As users cannot define their own "type keywords", the distinction is completely irrelevant from a user perspective.
2. I don't.

**Comment by soc** [comment]
@CyrusNajmabadi Agreed. Not sure what you are arguing against?

**Comment by CyrusNajmabadi** [comment]
> @CyrusNajmabadi Agreed. Not sure what you are arguing against?

You said to Youssef1313:

> @Youssef1313 The point is consistency: Do not make up rules for language users that you, as a language developer, don't intend to follow.

I am arguing against that idea.

The point is exactly to make up rules for users that the language itself won't follow.  We are trying to split this space *entirely* so that we can have more names like `int` and not have any chance of them colliding with users.

**Comment by soc** [comment]
Perhaps this issue and the discussion notes are misleading, because it's simply not "warn for lower-cased type names" then.

**Comment by CyrusNajmabadi** [comment]
>  because it's simply not "warn for lower-cased type names" then.

```
> During LDM discussion on records, there was interest to have an analyzer warn on lower-cased type names. 
```

And

https://github.com/dotnet/roslyn/pull/56905
`Report a level 7 warning for lower-cased type names`

It is exactly that.

**Comment by soc** [comment]
So when compiling the BCL we'll get warnings for `int`, `float`, `string`?

**Comment by CyrusNajmabadi** [comment]
> So when compiling the BCL we'll get warnings for int, float, string?

No.  The bcl does not define those.  So there's nothing to warn on.

**Comment by 333fred** [comment]
I think there might be a bit of confusion: `int`, `float`, `string`, and all the rest of the keywords on https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/built-in-types are not type names. They are C# keywords. The actual name for `string` is `System.String`, which does indeed follow this convention.

**Comment by soc** [comment]
Then "warn for lower-cased type names, except our own" or "warn for lower-cased type names, except for type keywords" would be correct, but the title as given is misleading.

**Comment by soc** [comment]
@333fred I'm not confused thank you.

**Comment by Youssef1313** [comment]
There is really no exception. `int` is a keyword. It's not a type name. The compiler warns for lower-cased type names.

**Comment by soc** [comment]
@Youssef1313 If you discuss type names then the "type-ness" of "type keywords" is more relevant that their "keyword-ness".

**Comment by CyrusNajmabadi** [comment]
> @CyrusNajmabadi Then "warn for lower-cased type names, except our own"

Those are not type-names.  They are built in keywords.  We're not warning on keywords here, only type names.  Thanks :)

> @Youssef1313 If you discuss type names then the "type-ness" of "type keywords" is more relevant that their "keyword-ness".

I'm not sure what point you're trying to make here.  We're warning on lowercase type names because that's the space that we want to reserve for us for keywords (type or otherwise).

**Comment by soc** [comment]
> I'm not sure what point you're trying to make here. 

Yes, that's the impression I got. Have a nice day!

**Comment by CyrusNajmabadi** [comment]
@soc  the relevance is the 'keyword-ness'.  We want to be able to make keywords that do not collide with user types.  It has nothing to do with the keyword being type-like (like `int`).  For example, even basic non-type keywords (like `record`, which doesn't correspond to any .net type) had a problem because it might collide with a user type.  I think you're not understanding the reason for this work, or how it relates to the language so far, or what we're considering in the future.

**Comment by davidwengier** [comment]
> Is there already a recommended time table on which users can start replacing int, float, string and friends?

If you or anyone else wants to disallow using language keywords like these in your own projects, you can enable this style rule and make it a warning or error, as you see fit: https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/style-rules/ide0049
